Fix font generation where width >=16 by reworking the chargen logic

DHowett opened the pull request:
This this pull request rewrites the character generator outer loop to be
based on which uint8 the character starts in, rather an an "index +
subunit." They're largely fungible, but there's a straightforward
transformation from u8 index to u32 position plus subunit.

The original code had an issue where it would preferentially store
glyphs in the 0- and 3- subunits when the font width was a multiple of 8
pixels. Because of that, it would cut off glyphs when the width was from 16
to 31. See Examples below for more info.

To simplify some of the logic, I also made it preallocate the entire character
data array so there's no more guesswork around whether to append or update.

For ease of understanding (and reproduceability!) I made it sort the
character map before inserting it into the packed bitset, and added a
huge block comment explaining how pixel font packing works.

This change in packing logic didn't require an equivalent change in the
font renderer, and I've tested and verified the generated fonts in a
couple applications.

## Examples

I ran a quick test with fonts of width 7, 8, 9, 16, 17 and 32. Glyph "A"
is a repeating "`XXX___`" pattern, glyph "B" is "`XX__XX__`" and glyph "C"
is "`X_X_X_`"

### Before
```
width_07.txt
_________CCCCCCC_BBBBBBB_AAAAAAA
00000000010101010011001101000111

width_08.txt
________CCCCCCCC________BBBBBBBB
00000000010101010000000000110011
________________________AAAAAAAA
00000000000000000000000011000111

width_09.txt
_______CCCCCCCCC_______BBBBBBBBB
00000001010101010000000100110011
_______________________AAAAAAAAA
00000000000000000000000111000111

width_16.txt
AAAAAAAA&&&&&&&&CCCCCCCCCCCCCCCC
11000111000000000101010101010101
________________BBBBBBBBBBBBBBBB
00000000000000000011001100110011

width_17.txt
BBBBBBBB&&&&&&&AAAAAAAAAAAAAAAAA
00110011000000000111000111000111
_______________CCCCCCCCCCCCCCCCC
00000000000000010101010101010101

width_32.txt
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
11000111000111000111000111000111
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
00110011001100110011001100110011
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
01010101010101010101010101010101
```

The bits marked with `&` were erroneously skipped, and the 16/17-width
fonts had glyphs A and B cut off (respectively). The 8-pixel-wide font
was packed sparsely (at approximately double the space it should
require)

### After

```
width_07.txt
_________CCCCCCC_BBBBBBB_AAAAAAA
00000000010101010011001101000111

width_08.txt
________CCCCCCCCBBBBBBBBAAAAAAAA
00000000010101010011001111000111

width_09.txt
_______BBBBBBBBB_______AAAAAAAAA
00000001001100110000000111000111
_______________________CCCCCCCCC
00000000000000000000000101010101

width_16.txt
BBBBBBBBBBBBBBBBAAAAAAAAAAAAAAAA
00110011001100110111000111000111
________________CCCCCCCCCCCCCCCC
00000000000000000101010101010101

width_17.txt
_______________AAAAAAAAAAAAAAAAA
00000000000000000111000111000111
_______________BBBBBBBBBBBBBBBBB
00000000000000010011001100110011
_______________CCCCCCCCCCCCCCCCC
00000000000000010101010101010101

width_32.txt
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
11000111000111000111000111000111
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
00110011001100110011001100110011
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
01010101010101010101010101010101
```

DHowett committed:
Fix font generation where width >=16 by reworking the chargen logic

This this pull request rewrites the character generator outer loop to be
based on which uint8 the character starts in, rather an an "index +
subunit." They're largely fungible, but there's a straightforward
transformation from u8idx to u32 position plus subunit.

The original code had an issue where it would preferentially store
glyphs in the 0- and 3- subunits when the font width was a multiple of 8
pixels. See Examples below for more info.

I also made it preallocate the entire character data array so we don't
need to conditionally append or update.

For ease of understanding (and reproduceability!) I made it sort the
character map before inserting it into the packed bitset, and added a
huge block comment explaining how pixel font packing works.

This change in packing logic didn't require an equivalent change in the
font renderer, and I've tested and verified the generated fonts in a
couple applications.

I ran a quick test with fonts of width 7, 8, 9, 16, 17 and 32. Glyph "A"
is a repeating "XXX___" pattern, glyph "B" is "XX__XX__" and glyph "C"
is "X_X_X_"

```
width_07.txt
_________CCCCCCC_BBBBBBB_AAAAAAA
00000000010101010011001101000111

width_08.txt
________CCCCCCCC________BBBBBBBB
00000000010101010000000000110011
________________________AAAAAAAA
00000000000000000000000011000111

width_09.txt
_______CCCCCCCCC_______BBBBBBBBB
00000001010101010000000100110011
_______________________AAAAAAAAA
00000000000000000000000111000111

width_16.txt
AAAAAAAA&&&&&&&&CCCCCCCCCCCCCCCC
11000111000000000101010101010101
________________BBBBBBBBBBBBBBBB
00000000000000000011001100110011

width_17.txt
BBBBBBBB&&&&&&&AAAAAAAAAAAAAAAAA
00110011000000000111000111000111
_______________CCCCCCCCCCCCCCCCC
00000000000000010101010101010101

width_32.txt
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
11000111000111000111000111000111
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
00110011001100110011001100110011
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
01010101010101010101010101010101
```

The bits marked with `&` were erroneously skipped, and the 16/17-width
fonts had glyphs A and B cut off (respectively). The 8-pixel-wide font
was packed sparsely (at approximately double the space it should
require)

```
width_07.txt
_________CCCCCCC_BBBBBBB_AAAAAAA
00000000010101010011001101000111

width_08.txt
________CCCCCCCCBBBBBBBBAAAAAAAA
00000000010101010011001111000111

width_09.txt
_______BBBBBBBBB_______AAAAAAAAA
00000001001100110000000111000111
_______________________CCCCCCCCC
00000000000000000000000101010101

width_16.txt
BBBBBBBBBBBBBBBBAAAAAAAAAAAAAAAA
00110011001100110111000111000111
________________CCCCCCCCCCCCCCCC
00000000000000000101010101010101

width_17.txt
_______________AAAAAAAAAAAAAAAAA
00000000000000000111000111000111
_______________BBBBBBBBBBBBBBBBB
00000000000000010011001100110011
_______________CCCCCCCCCCCCCCCCC
00000000000000010101010101010101

width_32.txt
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
11000111000111000111000111000111
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
00110011001100110011001100110011
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
01010101010101010101010101010101
```

DHowett commented:
I recognize that coming into somebody's project and rewriting some core logic is not often a winning strategy, so please feel free to reject this. I have a couple other fontgen improvements waiting in the wings, as well, but I'm also happy to maintain a fork of the font generator :smile:

pbnjay commented on cmd/fontgen/fontgen.go:
```diff
@@ -61,29 +62,67 @@ func generatePixFont(name string, w, h int, v bool, d map[rune]map[int]string) {
+	// 6                                E == 0b00000000000000000000000000000001 == 0x1
+	// 7                             EEEE == 0b00000000000000000000000000001111 == 0xf
+	// 8                                E == 0b00000000000000000000000000000001 == 0x1
+	// 9                            EEEEE == 0b00000000000000000000000000011111 == 0x1f
```
Can you pad the hex here so it also lines up?

pbnjay commented on cmd/fontgen/fontgen.go:
```diff
@@ -61,29 +62,67 @@ func generatePixFont(name string, w, h int, v bool, d map[rune]map[int]string) {
-		}
+	//
+	// This packed representation stores 1-4 glyphs in a single uint32 (per line).
+	// Each uint32 is trested as four individually-addressable uint8s, and each uint8
```
typo "treated" - also if we're going to have the description here, can you add a note that the renderer needs the byte-alignment for efficiency?

pbnjay commented on cmd/fontgen/fontgen.go:
```diff
@@ -61,29 +62,67 @@ func generatePixFont(name string, w, h int, v bool, d map[rune]map[int]string) {
+	// That is to say, a 4-character font with an 8-pixel width will be stored
+	// one u32 per line, and a 4-character font with a 9-pixel width will be
+	// stored 2 u32 per line (since you cannot fit >2 9-bit integers in a single
+	// u32 and maintain the 8-bit alignment guarantee)
```
"a 4-character font with an 8-pixel width will be stored one u32 per line" took me a bit to parse, I think 
"an 8-pixel width font can store 4 characters using one u32 per line"

and then "a 9-pixel width font can store 2 characters using one u32 per line since 9-bit values must be byte aligned" etc

pbnjay reviewed (commented):
Thanks!

Obviously this is some pretty old code! Looking back now, it's definitely overcomplicated and your code is similar to the approach I would have used to handle this bug.

I'm happy to merge this, the comments are good but I just made a few suggestions so that the next person who comes along has a chance of understanding it all!

I know there's no testing anywhere right now, but if your tests are native go tests feel free to include them so we can catch bugs like this in the future.

DHowett committed:
Address PR comments I

DHowett committed:
Update comment, pull pack out and test it

DHowett commented:
Alright, I reworked the big block comment. Even I was having trouble following it 😅.

I also pulled the packing logic out of `generatePixFont` so that it could be tested, and introduced some tests based on my original test files.

Because of the signature change, the incremental diff for the two most recent commits is probably going to be easier to follow than the entire PR diff!

DHowett commented:
```
=== RUN   TestGlyphPacking
=== RUN   TestGlyphPacking/7x1
=== RUN   TestGlyphPacking/8x1
=== RUN   TestGlyphPacking/9x1
=== RUN   TestGlyphPacking/16x1
=== RUN   TestGlyphPacking/17x1
=== RUN   TestGlyphPacking/32x1
--- PASS: TestGlyphPacking (0.00s)
    --- PASS: TestGlyphPacking/7x1 (0.00s)
    --- PASS: TestGlyphPacking/8x1 (0.00s)
    --- PASS: TestGlyphPacking/9x1 (0.00s)
    --- PASS: TestGlyphPacking/16x1 (0.00s)
    --- PASS: TestGlyphPacking/17x1 (0.00s)
    --- PASS: TestGlyphPacking/32x1 (0.00s)
PASS
ok      github.com/pbnjay/pixfont/cmd/fontgen   0.038s
```

😀

DHowett commented:
Ah, this isn't quite right yet. I added the little 5x5 font containing `ABCDE` to the test cases, and it miscalculates the number of u32s it'll need.

**EDIT**: This is fixed with a4b73b2

DHowett committed:
Add the 5x5 5-character font from the sample to the test

fix the cost per line calculation
